codingStandards/AdditionalGuidelines-Java.md

```diff
@@ -0,0 +1,98 @@
+# Additional Guidelines - Java
+
+## Argument validation
+
+Validating arguments in _every_ method can be an unnecessary cost.
+Following guidelines try to reduce the argument validation cost without losing its benefits. 
+
+1. **Validate an argument only if an invalid value can result in an _undesirable_ yet _non-failure_ behavior _directly_ 
+   within the method.**
+
+   In the example below, `name` should be validated because an invalid `name` can result in invalid data in 
+   the system from that point onwards without indicating that something went wrong.
+   
+   ```java
+   void setName(String name){
+       this.name = name;
+   }
+   ```
+   
+   However, `name` in the example below need not be validated because it is not being used _directly_ in the method.
+   
+   ```java
+   void setName(String name){
+       this.saveName(name);
+   }
+   ```
+   
+   In the example below, `key` need not be validated against `null` because a `null` value is going to create a 
+   detectable failure (i.e. a `NullPointerException`) anyway.
```

Comment:
> Java 7 and above actually has an Objects::requireNonNull() method which might be useful in this regard.

Didn't know such a thing existed :-p But it's not a replacement for assert? Probably it's more useful when we want to have argument checking in public APIs where asserts are not recommended.

Another related feature that might be relevant is annotations https://blogs.oracle.com/java-platform-group/entry/java_8_s_new_type Unfortunately, it's not supported in the standard jdk though.